simple_history/manager.py

```diff
@@ -9,13 +9,114 @@
+        - https://github.com/jazzband/django-simple-history/pull/229
+        - https://github.com/jazzband/django-simple-history/issues/354
+        - https://github.com/jazzband/django-simple-history/issues/397
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(HistoricalQuerySet, self).__init__(*args, **kwargs)
+        self._as_instances = False
+        self._pk_attr = self.model.instance_type._meta.pk.attname
+
+    def as_instances(self):
+        """
+        Return a queryset that generates instances instead of historical records.
+        Queries against the resulting queryset will translate `pk` into the
+        primary key field of the original type.
+
+        Returns a queryset.
+        """
+        if not self._as_instances:
+            result = self.exclude(history_type="-")
+            result._as_instances = True
+        else:
+            result = self._clone()
+        return result
+
+    def filter(self, *args, **kwargs):
+        """
+        If a `pk` filter arrives and the queryset is returning instances
+        then the caller actually wants to filter based on the original
+        type's primary key, and not the history_id (historical record's
+        primary key); this happens frequently with DRF.
+        """
+        if self._as_instances and "pk" in kwargs:
+            kwargs[self._pk_attr] = kwargs.pop("pk")
+        return super().filter(*args, **kwargs)
+
+    def latest_of_each(self):
+        """
+        Ensures results in the queryset are the latest historical record for each
+        primary key.  Deletions are not removed.
+
+        Returns a queryset.
+        """
+        # If using MySQL, need to get a list of IDs in memory and then use them for the
+        # second query.
+        # Does mean two loops through the DB to get the full set, but still a speed
+        # improvement.
+        backend = connection.vendor
+        if backend == "mysql":
+            history_ids = {}
+            for item in self.order_by("-history_date", "-pk"):
+                if getattr(item, self._pk_attr) not in history_ids:
+                    history_ids[getattr(item, self._pk_attr)] = item.pk
+            latest_historics = self.filter(history_id__in=history_ids.values())
+        elif backend == "postgresql":
+            latest_pk_attr_historic_ids = (
+                self.order_by(self._pk_attr, "-history_date", "-pk")
+                .distinct(self._pk_attr)
+                .values_list("pk", flat=True)
+            )
+            latest_historics = self.filter(history_id__in=latest_pk_attr_historic_ids)
+        else:
+            latest_pk_attr_historic_ids = (
+                self.filter(**{self._pk_attr: OuterRef(self._pk_attr)})
+                .order_by("-history_date", "-pk")
+                .values("pk")[:1]
+            )
+            latest_historics = self.filter(
+                history_id__in=Subquery(latest_pk_attr_historic_ids)
```

Comment:
I didn't realize this wasn't new code. You can ignore the following if you'd like.

Since you have an outer ref to `self._pk_attr` whih is doing association, you can have the following:

```
self.filter(Exists(latest_pk_attr_historic_ids))
```

Reply:
Most of the subquery code was simply moved into HistoricalQuerySet so it could be more selectively applied; and was not otherwise changed.